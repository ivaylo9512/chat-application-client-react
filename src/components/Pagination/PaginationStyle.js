import styled from "styled-components";

export const Container = styled.div`
    text-align: center;
`

export const Ul = styled.ul`
    text-align: center;
    margin: 0;
    padding: 0;
    display: inline-block;
`

export const Li = styled.li`
    list-style: none;
    display: inline-block;
    border-radius: 0.2vw;
    margin-right: 0.2vw;
    font-size: 0.6vw;
    padding: 0.6vw 1.3vw;
    border: 1px solid rgb(24,27,33);
    cursor: pointer;

    background: ${props => props.isSelected ? 'transparent' : 'rgb(24,27,33)'};
    color: ${props => props.isSelected ? 'rgb(24,27,33)' : 'white'};
    text-decoration: ${props => props.isSelected ? 'underline' : 'none'};
    box-shadow:  ${props => props.isSelected ? 'none' : '0px 0px 0.3vw 0px rgb(0 0 0 / 75%)'};
    border-color ${props => props.isSelected ? 'transparent' : 'rgb(24,27,33)'};


`