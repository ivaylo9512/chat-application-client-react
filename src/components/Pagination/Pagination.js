import { useDispatch, useSelector } from "react-redux";
import { Container, Ul, Li } from './PaginationStyle';

const Pagination = ({selector, setData, getData, pagesPerSlide = 4}) => {
    const { isLoading, dataInfo: { data, maxPages, currentPage, pages }, query } = useSelector(selector);
    const dispatch = useDispatch();

    const changePage = (page) => {
        if(isLoading){
            return;
        }

        const currentData = data[page - 1];
        if(currentData){
            return dispatch(setData({ currentData, currentPage: page }));
        }

        const pagesCount = page - pages; 
        dispatch(getData({...query, pages: pagesCount}))
    }


    return(
        <>
        {maxPages > 0 &&
            <Container>
                {currentPage > 1 &&
                    <button data-testid='back' onClick={() => changePage(currentPage - 1)}>prev</button>
                }
                <Ul>
                    {
                        Array.from({ length: currentPage / pagesPerSlide < 1 ? pagesPerSlide : pagesPerSlide + 1 }).map((el, i) => {
                            const slide = Math.floor(currentPage / pagesPerSlide);
                            const start = slide * pagesPerSlide + (slide == 0 ? 1 : 0);
                            const pageIndex = start + i;

                            if(pageIndex <= maxPages){
                                return <Li data-testid={`${pageIndex}`} pagesOnSlide={maxPages - start} isSelected={pageIndex == currentPage} key={pageIndex} onClick={() => changePage(pageIndex)}>{pageIndex}</Li>}
                            }
                        )
                    }
                </Ul>
                {currentPage < maxPages &&
                    <button data-testid='next' onClick={() => changePage(currentPage + 1)}>next</button>
                }
            </Container>
            }
        </>
    )
}
export default Pagination