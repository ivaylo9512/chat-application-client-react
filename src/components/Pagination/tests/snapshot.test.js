import { shallow } from 'enzyme';
import Pagination from 'components/Pagination/Pagination';
import * as Redux from 'react-redux';

describe('Pagination snapshot tests', () => {
    let selectorSpy;

    beforeEach(() => {
        selectorSpy = jest.spyOn(Redux, 'useSelector');
        jest.spyOn(Redux, 'useDispatch').mockReturnValue(jest.fn());
    })

    const createWrapper = (state) => {
        selectorSpy.mockReturnValue(state);

        return shallow(<Pagination data={[]} getData={jest.fn()} pagesPerSlide={5}/>)
    }

    it('renders correctly with pages', () => {
        const wrapper = createWrapper({ data: { maxPages: 5, isLoading: false}, query: { name: '', take: '10'}});

        expect(wrapper).toMatchSnapshot();
    })

    it('renders correctly without pages', () => {
        const wrapper = createWrapper({ data: { maxPages: 0, isLoading: false}, query: { name: '', take: '10'}});

        expect(wrapper).toMatchSnapshot();
    })

    it('renders correctly with pages at page 2 and back button', () => {
        const wrapper = createWrapper({ data: { maxPages: 5, isLoading: false}, query: { name: '', take: '10'}});
        wrapper.findByTestid(2).simulate('click');

        expect(wrapper).toMatchSnapshot();
    })

    it('renders correctly with page that is last page of slide should render next slide', () => {
        const wrapper = createWrapper({ data: { maxPages: 10, isLoading: false}, query: { name: '', take: '10'}});
        wrapper.findByTestid(5).simulate('click');

        expect(wrapper).toMatchSnapshot();
    })

    it('renders correctly with pages at last page', () => {
        const wrapper = createWrapper({ data: { properties: [], maxPages: 4, isLoading: false}, query: { name: '', take: '10'}});
        wrapper.findByTestid(4).simulate('click');

        expect(wrapper).toMatchSnapshot();
    })

})