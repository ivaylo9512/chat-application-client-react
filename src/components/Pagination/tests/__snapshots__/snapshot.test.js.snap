// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination snapshot tests renders correctly with page that is last page of slide should render next slide 1`] = `
<Fragment>
  <styled.div>
    <button
      data-testid="back"
      onClick={[Function]}
    >
      prev
    </button>
    <styled.ul>
      <styled.li
        data-testid="5"
        isSelected={true}
        key="5"
        onClick={[Function]}
        pagesOnSlide={5}
      >
        5
      </styled.li>
      <styled.li
        data-testid="6"
        isSelected={false}
        key="6"
        onClick={[Function]}
        pagesOnSlide={5}
      >
        6
      </styled.li>
      <styled.li
        data-testid="7"
        isSelected={false}
        key="7"
        onClick={[Function]}
        pagesOnSlide={5}
      >
        7
      </styled.li>
      <styled.li
        data-testid="8"
        isSelected={false}
        key="8"
        onClick={[Function]}
        pagesOnSlide={5}
      >
        8
      </styled.li>
      <styled.li
        data-testid="9"
        isSelected={false}
        key="9"
        onClick={[Function]}
        pagesOnSlide={5}
      >
        9
      </styled.li>
      <styled.li
        data-testid="10"
        isSelected={false}
        key="10"
        onClick={[Function]}
        pagesOnSlide={5}
      >
        10
      </styled.li>
    </styled.ul>
    <button
      data-testid="next"
      onClick={[Function]}
    >
      next
    </button>
  </styled.div>
</Fragment>
`;

exports[`Pagination snapshot tests renders correctly with pages 1`] = `
<Fragment>
  <styled.div>
    <styled.ul>
      <styled.li
        data-testid="1"
        isSelected={true}
        key="1"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        1
      </styled.li>
      <styled.li
        data-testid="2"
        isSelected={false}
        key="2"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        2
      </styled.li>
      <styled.li
        data-testid="3"
        isSelected={false}
        key="3"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        3
      </styled.li>
      <styled.li
        data-testid="4"
        isSelected={false}
        key="4"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        4
      </styled.li>
      <styled.li
        data-testid="5"
        isSelected={false}
        key="5"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        5
      </styled.li>
    </styled.ul>
    <button
      data-testid="next"
      onClick={[Function]}
    >
      next
    </button>
  </styled.div>
</Fragment>
`;

exports[`Pagination snapshot tests renders correctly with pages at last page 1`] = `
<Fragment>
  <styled.div>
    <button
      data-testid="back"
      onClick={[Function]}
    >
      prev
    </button>
    <styled.ul>
      <styled.li
        data-testid="1"
        isSelected={false}
        key="1"
        onClick={[Function]}
        pagesOnSlide={3}
      >
        1
      </styled.li>
      <styled.li
        data-testid="2"
        isSelected={false}
        key="2"
        onClick={[Function]}
        pagesOnSlide={3}
      >
        2
      </styled.li>
      <styled.li
        data-testid="3"
        isSelected={false}
        key="3"
        onClick={[Function]}
        pagesOnSlide={3}
      >
        3
      </styled.li>
      <styled.li
        data-testid="4"
        isSelected={true}
        key="4"
        onClick={[Function]}
        pagesOnSlide={3}
      >
        4
      </styled.li>
    </styled.ul>
  </styled.div>
</Fragment>
`;

exports[`Pagination snapshot tests renders correctly with pages at page 2 and back button 1`] = `
<Fragment>
  <styled.div>
    <button
      data-testid="back"
      onClick={[Function]}
    >
      prev
    </button>
    <styled.ul>
      <styled.li
        data-testid="1"
        isSelected={false}
        key="1"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        1
      </styled.li>
      <styled.li
        data-testid="2"
        isSelected={true}
        key="2"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        2
      </styled.li>
      <styled.li
        data-testid="3"
        isSelected={false}
        key="3"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        3
      </styled.li>
      <styled.li
        data-testid="4"
        isSelected={false}
        key="4"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        4
      </styled.li>
      <styled.li
        data-testid="5"
        isSelected={false}
        key="5"
        onClick={[Function]}
        pagesOnSlide={4}
      >
        5
      </styled.li>
    </styled.ul>
    <button
      data-testid="next"
      onClick={[Function]}
    >
      next
    </button>
  </styled.div>
</Fragment>
`;

exports[`Pagination snapshot tests renders correctly without pages 1`] = `<Fragment />`;
